Assignment2Q2.java

```diff
@@ -0,0 +1,129 @@
+		BinaryTreeNode<Integer> n5 = new BinaryTreeNode<>(9, n3, n4);
+		BinaryTreeNode<Integer> n6 = new BinaryTreeNode<>(19, null, null);
+		BinaryTreeNode<Integer> n7 = new BinaryTreeNode<>(18, null, n6);
+		BinaryTreeNode<Integer> n8 = new BinaryTreeNode<>(16, n5, n7);
+		BinaryTreeNode<Integer> n9 = new BinaryTreeNode<>(186, null, null);
+
+		System.out.println(findCommonAncestors(n9, n1, n8).getData());
+	}
+
+	/**
+	 * Finds lowest common ancestor of node1 and node2
+	 * 
+	 * @param node1
+	 *            first node
+	 * @param node2
+	 *            second node
+	 * @param root
+	 *            root of the tree
+	 * @return lowest common ancestor of node1 and node2
+	 * @throws IllegalArgumentException
+	 *             if one or both of the nodes are not present in the tree
+	 */
+	public static <T> BinaryTreeNode<T> findCommonAncestors(BinaryTreeNode<T> node1, BinaryTreeNode<T> node2,
+			BinaryTreeNode<T> root) throws IllegalArgumentException {
+		if (node1 == null || node2 == null || root == null) {
+			throw new IllegalArgumentException();
+		}
+
+		BinaryTreeNode<T> node = findCommonAncestorsHelper(node1, node2, root);
+
+		if (node == null) {
+			throw new IllegalArgumentException();
+		}
+
+		// if node1 is the common ancestor, check if node2 is in the tree which
+		// has root node1. Also handles the case if node1 and node2 are the
+		// same.
+		if (node.getData().equals(node1.getData())) {
+			if (node1.getData().equals(node2.getData()) || contains(node2, node)) {
+				return node;
+			}
+
+			throw new IllegalArgumentException();
+		}
+
+		// if node2 is the common ancestor, check if node1 is in the tree which
+		// has root node2.
+		if (node.getData().equals(node2.getData())) {
+			if (contains(node1, node)) {
+				return node;
+			}
+
+			throw new IllegalArgumentException();
+		}
+
+		return node;
+	}
+
+	/**
+	 * Helper method of findCommonAncestors. Partially finds the common ancestor
+	 * of node1 and node2
+	 * 
+	 * @param node1
+	 *            first node
+	 * @param node2
+	 *            second node
+	 * @param root
+	 *            root of the tree
+	 * @return common ancestor of node1 and node2
+	 */
+	private static <T> BinaryTreeNode<T> findCommonAncestorsHelper(BinaryTreeNode<T> node1, BinaryTreeNode<T> node2,
+			BinaryTreeNode<T> root) {
+		if (root == null) {
+			return null;
+		}
+
+		if (root.getData().equals(node1.getData()) || root.getData().equals(node2.getData())) {
+			return root;
+		}
+
+		// common ancestor of left subtree
+		BinaryTreeNode<T> leftCA = findCommonAncestorsHelper(node1, node2, root.getLeft());
+		// common ancestor of right subtree
+		BinaryTreeNode<T> rightCA = findCommonAncestorsHelper(node1, node2, root.getRight());
+
+		// one of the nodes is on the left subtree, other is on the right
+		// subtree
+		if (leftCA != null && rightCA != null) {
+			return root;
+		}
+
+		// both are on the left sub tree
+		if (leftCA != null) {
+			return leftCA;
+		}
+
+		// both are on the right subtree

```

Comment:
Is just changing the comment to "both are on the right subtree or key does not exist in the tree" ok or should I change the code too, to make it more readable?